client/src/components/Gallery/Gallery.js

```diff
@@ -1,17 +1,59 @@
 import React from 'react';
 import './Gallery.scss';
-import { Modal, Pagination } from 'react-bootstrap';
+import { Modal } from 'react-bootstrap';
 
 export default class Gallery extends React.Component {
 
     constructor(props) {
         super(props);
+        let selectedImageIndex = props.selectedImageIndex || 0;
+        if (props.images && props.images.length) {
+            selectedImageIndex = props.images.findIndex(image => image.isPrimary) || selectedImageIndex;
+        }
+        const imageRotationAnimationLength = props.imageRotationAnimationLength || 1000;
+        const imageRotationInterval = (props.imageRotationInterval || 10000) - imageRotationAnimationLength;
         this.state = {
             modalOpen: false,
-            modalPage: 0
+            modalPage: 0,
+            selectedImageIndex,
+            imageRotationInterval,
+            imageRotationAnimationLength,
+            animating: false,
+            imageTimeout: window.setTimeout(() => this.nextImage(), imageRotationInterval)
         };
     }
 
+    nextImage() {
+        let index = this.state.selectedImageIndex;
+        index++;
+        if (index === this.props.images.length) index = 0;
+        const previousIndex = this.state.selectedImageIndex;
+        this.animate(index, previousIndex);
+    }
+
+    selectImage(index) {
+        window.clearTimeout(this.state.imageTimeout);
+        const previousIndex = this.state.selectedImageIndex;
+        this.animate(index, previousIndex);
+    }
+
+    animate(index, previousIndex) {
+        this.setState({
+            animationIndex: index,
+            animating: true
+        });
+        const { imageTimeout, imageRotationAnimationLength } = this.state;
+        window.setTimeout(() => {
```

Comment:
Can you explain what this fancy code is doing?

Reply:
I can try: basically, there's 2 image divs, one sits on top of the other covering it. The bottom div is our "real" image, the top one is the one used solely by the animation. The way the animation works is basically:

No animation is running
Image A is in bottom div
Top div is empty and not showing

Then, we change to Image B and this happens:
Animation is running
Image B is placed in bottom div
Image A is placed in top div with an opacity of 1, slowly animating to opacity 0, in other words disappearing

Animation is complete
Image B is still in the bottom div
The top div is now made fully invisible (due to its black background, it can't be visible while not animating)

The javascript that makes this work basically runs on a 1 second timer which is what setTimeout does. There are a lot of nuances to it that are easier to explain in person.

Reply:
I also just added even more complexity to this by delaying part of the last step 100ms to give the image time to load in, otherwise you would get a weird flashing glitch as the image popped into a new div